### ISSUE TITLE
Zero length segments are not errors in Inkscape but are in Inkstitch

### ISSUE THREAD

**Issue by jhughes2112** [inkstitch/inkstitch#1267]
It is common for an image trace to generate zero length segments on paths, and unfortunately this is not considered an error in Inkscape: https://www.w3.org/TR/SVG/paths.html#ZeroLengthSegments

So far this has happened on about 1 out of every 4 or 5 traces we've done, and today we finally figured out what the problem was.  Still looking for a more automated way to handle it, but closely scanning the nodes with your naked eye will show them rendering differently if there are coincident nodes.  There's some magical function (shift-J for join selected nodes) that is supposed to help, but I haven't gotten it to behave properly yet, and haven't found any "select zero length segments" function to make this easier.

However, this causes a lot of errors that prevent images from properly rendering stitches.  You can replicate this rather easily by dragging a node along a segment directly on top of another, then go to Inkscape->Params.

I would suggest one of the following:
1) Fix it on your end to ignore zero length segments, since that's valid input.
2) Select whatever nodes are triggering the error upon failure, zoom in on the offending edge, or SOME other useful user feedback mechanism other than "path 24770: error: fill: shape is not valid. This can happen if the border crosses over itself."  I have not found any mechanism for selecting a path by index, so this is not particularly useful.

Thanks so much for this really helpful tool!  It's giving us a ton of joy over here!

**Comment by lexelby** [comment]
Thanks for reporting this!  I'm definitely in favor of option 1.

Do you always see the "not valid" message?  Or do you ever see a crash with stack trace?  If you get a stack trace, posting it here as a comment would be really helpful.

I've actually found a specific use case for zero-length paths in Ink/Stitch in the past.  My first embroidery machine was an entry-level Brother SE400 that had no jump-stitch trim functionality.  Often, I would have an extremely short jump stitch that was hard to trim manually (example: between letters in text).

I experimented with creating a single-node path with a start marker so that I could see it.  I placed this marker off to the side, causing a single stitch to be made outside my design between each letter.

This gave me a nice long length of thread to hold onto when trimming.  I would snip the bobbin thread, releasing the long stitch, then hold it on top while I trimmed it away.  Very useful.

I think I may have had to modify Ink/Stitch to make this work.  Advantages of being the developer. ;)

Bottom line: we definitely should support zero-length paths and path segments.  The fact that we don't is a bug.

**Comment by jhughes2112** [comment]
As far as I have seen, it has 100% of the time given the above error report without a call stack.  Given that it's a zero length segment, it's a special case that you probably can detect, but is currently considered an overlapping border by the code.

Interesting use for a jump stitch!  I hadn't thought of that before.  I was curious if there's a way to know what order the paths are converted to stitches?

I did find that you can use the Join Selected Nodes button to merge nodes, and that will get rid of the zero length segments, so there's a workaround, but it took a while to figure that out.  Lots of users may get discouraged before figuring that out.  One option may be to call that under the hood and undo it, so it's non-destructive.

**Comment by kaalleen** [comment]
#1230 is meant to fix this issue. It will not ignore small/empty fill elements, but instead it stitch them out. Troubleshoot will point on them to inform about their existence. If you want to remove them, you could use the cleanup document function in the troubleshoot section. Sorry for my brevity, I do not have a lot of time right now. Just wanted to give a short input. Hope it helps.

**Comment by lexelby** [comment]
As to the order objects stitch out, it's the same as the stacking order in the SVG, with lower down objects coming before higher up ones.  The Objects panel is really useful for visualizing the order.
